cbits/dns.c

```diff
@@ -3,74 +3,66 @@
+// error code. See: https://msdn.microsoft.com/en-us/library/windows/desktop/ms681382(v=vs.85).aspx
+DWORD getWindowsDefDnsServers(char* dnsAddresses, size_t bufferLen) {
     FIXED_INFO *pFixedInfo;
     ULONG ulOutBufLen;
     DWORD dwRetVal;
 
-    dns_t* result = (dns_t*) MALLOC(sizeof(dns_t));
-    if (result == NULL) return (dns_t*) ERROR_NOT_ENOUGH_MEMORY;
+    if (bufferLen <= 0) return ERROR_NOT_ENOUGH_MEMORY;
 
-    result->dnsAddresses = (char*) MALLOC(128 * sizeof(char));
-    if (result->dnsAddresses == NULL) {
-      result->error = ERROR_NOT_ENOUGH_MEMORY;
-      return result;
-    }
-
-    result->error = NO_ERROR;
     pFixedInfo = (FIXED_INFO *) MALLOC(sizeof (FIXED_INFO));
-    if (pFixedInfo == NULL) {
-      result->error = ERROR_NOT_ENOUGH_MEMORY;
-      return result;
-    }
+    if (pFixedInfo == NULL)
+      return ERROR_NOT_ENOUGH_MEMORY;
     ulOutBufLen = sizeof (FIXED_INFO);
 
     // Make an initial call to GetAdaptersInfo to get the necessary size into the
     // ulOutBufLen variable
     if (GetNetworkParams(pFixedInfo, &ulOutBufLen) == ERROR_BUFFER_OVERFLOW) {
         FREE(pFixedInfo);
         pFixedInfo = (FIXED_INFO *) MALLOC(ulOutBufLen);
-        if (pFixedInfo == NULL) {
-            result->error = ERROR_NOT_ENOUGH_MEMORY;
-            return result;
-        }
+        if (pFixedInfo == NULL)
+            return ERROR_NOT_ENOUGH_MEMORY;
     }
 
     dwRetVal = GetNetworkParams(pFixedInfo, &ulOutBufLen);
 
     if (dwRetVal == NO_ERROR) {
       int offset = 0;
-      int space_available = 255;
+      int spaceAvailable = bufferLen;
       IP_ADDR_STRING* head = &pFixedInfo->DnsServerList;
-      while (head != NULL && space_available >= 16) {
-        int ip_len = strlen(head->IpAddress.String);
-        strcpy_s(result->dnsAddresses + offset, ip_len + 1, head->IpAddress.String);
-        // Update the offset (non-null-terminated IP address + separator)
-        offset += ip_len + 1;
-        // Update the space available, pessimistically
-        space_available -= 16;
+
+      while (head != NULL && spaceAvailable >= 16) {
+        int ipLen = strlen(head->IpAddress.String);
+        int copySize = ipLen + 1;
+
+        // Copy the IP address, including the null terminator.
+        strcpy_s(dnsAddresses + offset, copySize, head->IpAddress.String);
+
+        spaceAvailable -= copySize;
+        if (spaceAvailable <= 0) break;
+
+        offset += copySize;
+
         // Write the separator, but only if this is not the last one,
-        // otherwise terminate the string.
+        // otherwise the string is already terminated due to the call to
+        // strcpy_s, which copies the null terminator.
+
         head = head->Next;
-        if (head == NULL)
-          result->dnsAddresses[offset] = '\0';
-        else
-          result->dnsAddresses[offset - 1] = ',';
+        if (head != NULL)
+          dnsAddresses[offset - 1] = ',';
```

Comment:
I spent a bit of time thinking of an edge case that continues causing the code to misbehave, and I think I found it.

Let's say that just before the program enters this line (line 58, `if (head != NULL)`), we have:
* `head != NULL`, so line 59 (`dnsAddresses[offset - 1] = ',';`) is executed
* `space_available < 16`

From this it follows that the `while` cycle will not run again, so `dnsAddresses` was modified such that its final character is `,` and not `\0` as intended. This may cause issues Haskell-side.

Reply:
To fix this,I suggest the following change to line 58:

`if (head != NULL)` becomes `if ((head != NULL) || (space_available >= 16))`.

Reply:
Alternatively, line 49 could change from `if (spaceAvailable <= 0) break;` to `if (spaceAvailable < 8) break;`.
The smallest possible IPv4 address is of the form "X.X.X.X", so 7 characters plus 1 null terminator = 8 characters. If the buffer can't hold at least 8 characters the `while` can be broken out of.

Reply:
I will fix this by myself.

Reply:
To remove the magic number, how about this?
```c
      int count = 0;

      while (head != NULL) {
        int ipLen = strlen(head->IpAddress.String);
        int copySize = ipLen + 1;

        spaceAvailable -= copySize;
	if (spaceAvailable >= 0) {
	  // Write the separator.
	  // The string is already terminated due to the call to
	  // strcpy_s, which copies the null terminator.
	  if (count != 0) dnsAddresses[offset - 1] = ',';
	  // Copy the IP address, including the null terminator.
	  strcpy_s(dnsAddresses + offset, copySize, head->IpAddress.String);
	} else
	  break;

        offset += copySize;
	count++;
        head = head->Next;
      }
```

Reply:
Note that 8 is not good enough for IPv6 addresses.

Reply:
@kazu-yamamoto I didn't consider that the functions exposed by `iphlpapi.h` may in the future return IPv6 addresses. You're right.

Regarding the C code you wrote, I think the problem persists.

Let's say that there are *at least* two IP addresses in the linked list. 
* If we only have one then the line `count != 0` won't ever be executed, and no commas will be added.
* After the `while` loop starts for the second address, if `space_available >= 0` holds (it will, after only 2 addresses and `bufferLen == 128`) we'll add a comma to the end of `dnsAddresses`, and then the `while` will be over. Once more, missing `\0`.

I feel like I'm bike-shedding at this point, so I apologize for that.

Reply:
I'm sorry but I don't understand your point.
When `space_available >= 0` holds, a comma is added AND `strcpy_s` is called.
So, `\0` is copied by `strcpy_s` in the last position.

Reply:
Yes exactly, it's important to recall that every call to `strcpy_s` will add `\0`, so if we guarantee to call it, and we don't do stupid things with the comma, we will be on the safe side I think.

Reply:
See my topic branch: https://github.com/kazu-yamamoto/dns/tree/input-output-hk-win-improvements

Reply:
Should we add `if (copySize == 0) continue;` instead?

Reply:
Never mind. This does not work because `ipLen + 1`.

So, let's add `if (spaceAvailable == 0) break;`.

Reply:
@kazu-yamamoto I looked at it closely and you're right - `strcpy_s` is always the last operation called that modifies `dnsAddresses`, so the `'\0'` will be present no matter what.